lib/src/Base/Solver/Fehlberg.cxx

```diff
@@ -0,0 +1,298 @@
+  // Quick return if the time grid is empty
+  const UnsignedInteger steps = timeGrid.getSize();
+  const UnsignedInteger dimension = transitionFunction_.getOutputDimension();
+  if (steps == 0) return Sample(steps, dimension);
+  // First, solve the ODE up to the final time and store all the intermediate values
+  Point state(initialState);
+  Scalar t = timeGrid[0];
+  Sample values(1, state);
+  Sample derivatives(0, dimension);
+  Point times(1, t);
+  Scalar h = ResourceMap::GetAsScalar("Fehlberg-InitialStep");
+  const Scalar tEnd = timeGrid[steps - 1];
+  const Bool positiveStep = tEnd > t;
+  if (!positiveStep) h = -h;
+  Bool done = false;
+  Point gradient;
+  while (!done)
+    {
+      Scalar newT = t + h;
+      if (positiveStep && (newT > tEnd) || !positiveStep && (newT < tEnd))
+	{
+	  done = true;
+	  h = tEnd - t;
+	  newT = tEnd;
+	}
+      state = computeStep(t, state, gradient, h);
+      values.add(state);
+      derivatives.add(gradient);
+      times.add(newT);
+      t = newT;
+    }
+  // Final evaluation of the gradient
+  derivatives.add(transitionFunction_(t, state));
+  // Now we interpolate the solution on the expected grid
+  PiecewiseHermiteEvaluation hermite(times, values, derivatives);
+  Sample result(steps, dimension);
+  for (UnsignedInteger i = 0; i < steps; ++i)
+    result[i] = hermite(Point(1, timeGrid[i]));
+  return result;
+}
+
+/* Perform one step of the Fehlberg method
+   See J. Stoer, R. Bulirsch, "Introduction to Numerical Analysis 2nd Edition", pp448-458.
+ */
+Point Fehlberg::computeStep(const Scalar t,
+			    const Point & state,
+			    Point & gradient,
+			    Scalar & h) const
+{
+  const UnsignedInteger dimension = state.getDimension();
+  Sample f(order_ + 2, dimension);
+  gradient = transitionFunction_(t, state);
+  f[0] = gradient;
+  UnsignedInteger index = 0;
+  for (UnsignedInteger k = 0; k <= order_; ++k)
+    {
+      const Scalar tK = t + alpha_[k] * h;
+      Point yK(state);
+      for (UnsignedInteger l = 0; l <= k; ++l)
+	{
+	  yK += f[l] * (h * beta_[index]);
+	  ++index;
+	}
+      f[k + 1] = transitionFunction_(tK, yK);
+    }
+  Point PhiI(dimension);
+  Point PhiII(dimension);
+  for (UnsignedInteger k = 0; k <= order_; ++k)
+    {
+      PhiI  += f[k] * c_[k];
+      PhiII += f[k] * cHat_[k];
+    }
+  PhiII += f[order_ + 1] * cHat_[order_ + 1];
+  const Point value(state + PhiII * h);
+  // The update formula for h is based on relation 7.2.5.17 in the reference,
+  // with the remark that \bar{y}_{i+1}-\hat{y}_{i+1}=h_i(\Phi_I-\Phi_{II})
+  const Scalar delta = (PhiI - PhiII).norm();
+  if (delta > 0.0)
+    {
+      const Scalar factor = std::exp(std::log(localPrecision_ / delta) / (order_ + 1.0));
+      h *= factor;
+    }
+  return value;
+}
+
+/* Method save() stores the object through the StorageManager */
+void Fehlberg::save(Advocate & adv) const
+{
+  ODESolverImplementation::save(adv);
```

Comment:
looks like some attributes should be serialized

Reply:
@jschueller Indeed...